src/tapedfunction.jl

```
@@ -168,127 +156,184 @@ function (instr::Instruction{F})() where F
+    ins = GotoInstruction(cond, line.dest + 1, taped)
+    push!(taped, ins)
+end
+
+function (builder::InstructionBuilder)(var::IRVar, line::Core.GotoNode)
+    (; taped, ir, _box) = builder
+    cond = Box{Any}(false) # unify the condiftion type
+    ins = GotoInstruction(cond, line.label + 1, taped)
+    push!(taped, ins)
+end
+
+function (builder::InstructionBuilder)(var::IRVar, line::Core.ReturnNode)
+    (; taped, ir, _box) = builder
+    ins = ReturnInstruction(_box(line.val), taped)
+    push!(taped, ins)
+end
+
+function (builder::InstructionBuilder)(var::IRVar, line::Expr)
+    builder(var, line, Val(line.head))
+end
+
+function (builder::InstructionBuilder)(var::IRVar, line::Expr, ::Val{:new})
+    (; taped, ir, _box) = builder
+    args = map(_box, line.args)
+    ins = Instruction(__new__, args |> Tuple, _box(var), taped)
+    push!(taped, ins)
+end
+
+function (builder::InstructionBuilder)(var::IRVar, line::Expr, ::Val{:call})
+    (; taped, ir, _box) = builder
+    args = map(_box, line.args)
+    # args[1] is the function
+    func = args[1]
+    if Meta.isexpr(func, :static_parameter) # func is a type parameter
+        func = ir.parent.sparam_vals[func.args[1]]
+    end
+    ins = Instruction(func, args[2:end] |> Tuple, _box(var), taped)
+    push!(taped, ins)
+end
+
+function (builder::InstructionBuilder)(var::IRVar, line::Expr, ::Val{:(=)})
+    (; taped, ir, _box) = builder
+    # args[1] (the left hand) is a SlotNumber, and it should be the output
+    rh = line.args[2] # the right hand, maybe a Expr, or a var, or ...
+    if Meta.isexpr(rh, [:new, :call])
+        builder(line.args[1], rh, Val(rh.head))
+    else # rh is a single value
+        ins = Instruction(identity, (_box(rh),), _box(line.args[1]), taped)
+        push!(taped, ins)
+    end
+end
+
+function (builder::InstructionBuilder)(var, line)
+    @error "Unknown IR code: " typeof(var) var typeof(line) line
+end
```

This will throw an error anyways:
```suggestion
```

---

But printing out more information on error is much useful to fix it. Without this, we only get a no-method matched error.

---

True. It would better to throw an error then and not only print an error message though.